nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/QueryDatabaseTable.java

```diff
@@ -0,0 +1,607 @@
+     * the wildcard (*) is used to select all columns.
+     *
+     * @param tableName   The name of the table to select from
+     * @param columnNames A comma-separated list of column names to select from the table
+     * @return A SQL select statement representing a query of the given column names from the given table
+     */
+    protected StringBuilder getSelectFromClause(String tableName, String columnNames) {
+        final StringBuilder query = new StringBuilder("SELECT ");
+        if (StringUtils.isEmpty(columnNames) || columnNames.trim().equals("*")) {
+            query.append("*");
+        } else {
+            query.append(columnNames);
+        }
+        query.append(" FROM ");
+        query.append(tableName);
+        return query;
+    }
+
+    /**
+     * Returns a SQL literal for the given value based on its type. For example, values of character type need to be enclosed
+     * in single quotes, whereas values of numeric type should not be.
+     *
+     * @param type  The JDBC type for the desired literal
+     * @param value The value to be converted to a SQL literal
+     * @return A String representing the given value as a literal of the given type
+     */
+    protected String getLiteralByType(int type, String value, String preProcessStrategy) {
+        // Format value based on column type. For example, strings and timestamps need to be quoted
+        switch (type) {
+            // For string-represented values, put in single quotes
+            case CHAR:
+            case LONGNVARCHAR:
+            case LONGVARCHAR:
+            case NCHAR:
+            case NVARCHAR:
+            case VARCHAR:
+            case ROWID:
+            case DATE:
+            case TIME:
+                return "'" + value + "'";
+            case TIMESTAMP:
+                // Timestamp literals in Oracle need to be cast with TO_DATE
+                if ("Oracle".equals(preProcessStrategy)) {
+                    return "to_date('" + value + "', 'yyyy-mm-dd HH24:MI:SS')";
+                } else {
+                    return "'" + value + "'";
+                }
+                // Else leave as is (numeric types, e.g.)
+            default:
+                return value;
+        }
+    }
+
+    protected class MaxValueResultSetRowCollector implements JdbcCommon.ResultSetRowCallback {
+        StateManager stateManager;
+        String preProcessStrategy;
+
+        public MaxValueResultSetRowCollector(StateManager stateManager, String preProcessStrategy) {
+            this.stateManager = stateManager;
+            this.preProcessStrategy = preProcessStrategy;
+        }
+
+        @Override
+        public void processRow(ResultSet resultSet) throws IOException {
+            if (resultSet == null) {
+                return;
+            }
+            try {
+                // Iterate over the row, check-and-set max values
+                StateMap stateMap = stateManager.getState(Scope.CLUSTER);
```

Comment:
This is going to make a call to ZooKeeper, generally, for every row that is selected. Need to pull this out, I think, so that it's done at most 1 per ResultSet. But since the onTrigger() method is also calling getState, I'd try to avoid doing it at all and just having this object created within onTrigger each time, passing in the StateMap to the constructor.